### PR TITLE
Filter_odd_numbers

### LINKED ISSUE #34: Print Odd Numbers in a List
# Odd Numbers Challenge

## Description:
Write a Python function that takes a list of integers and returns a new list containing only the odd numbers from the original list.

## Requirements:
- The function should return a list of integers.
- If the input list is empty, return an empty list.
- Only odd numbers should be included in the result.

### Constraints:
- The input list can contain negative integers. These should be treated the same as positive integers when checking for odd numbers.
- If a floating-point number is included in the input, it should be ignored in the output.
- Non-integer inputs are not valid and should be ignored without raising an error.

#### Valid Input:
- **Input**: `[1, 2, -3, 4.5, 5]`
- **Output**: `[1, -3, 5]`

#### Invalid Input (ignored):
- **Input**: `[1, 2, "three", 4.5, None]`
- **Output**: `[1]`

## Helpful Resources:
- [Python Lists](https://docs.python.org/3/tutorial/datastructures.html)
- [Python List Comprehensions](https://docs.python.org/3/tutorial/datastructures.html#list-comprehensions)
- [Python Type Checking](https://docs.python.org/3/library/typing.html)

### PR DESCRIPTION
**Pull request by Shahroz657**
fixes #34 
--- 
name: solution review 
about: A template PR for code review with a checklist
---

# Behavior
The function filters a list of integers, returning only the odd numbers. It ignores non-integer and floating-point inputs, handles negative integers, and returns an empty list for invalid or empty input.

## Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/filter_odd_numbers.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        >>> filter_odd_numbers([1, 2, "three", 4.5, None])
+        [1]
+
+        >>> filter_odd_numbers([])
+        []
+    """
+    assert isinstance(input_list, list), "Input must be a list."
+
+    # Filter and return only odd integers
+    return [x for x in input_list if isinstance(x, int) and x % 2 != 0]
```
Comment (MadiMalik) [inline solutions/filter_odd_numbers.py]: Great work on the functionality! However, this code could be simplified for better readability. While the list comprehension works well, separating the logic into smaller steps would be helpful

Comment (Shahroz657) [reply solutions/filter_odd_numbers.py]: Thanks for the feedback Madiha. I will certainly look into it

**Comment by MadiMalik** [comment]
Hey Shahroz,

Your code and tests are great, but there’s room for improvement. You can simplify your code by breaking it down into smaller, more straightforward steps. Additionally, consider adding test cases for boundary scenarios to make your tests even more better.

**Comment by Shahroz657** [comment]
"Hey Madiha,  
Thank you for your feedback! I appreciate your suggestions and will definitely look into simplifying the code further by breaking it into smaller, more straightforward steps.

Regarding the existing test cases and assertions:

1. **Defensive Assertions**:  
   - The function includes an assertion to ensure only lists are accepted as input:
     ```python
     assert isinstance(input_list, list), "Input must be a list."
     ```  
   - This is explicitly tested in the test suite with invalid inputs like:
     - **String**: `"not a list"`
     - **Integer**: `12345`
     - **Dictionary**: `{"key": "value"}`
     - **None**: `None`  

   These tests confirm that the function raises an `AssertionError` for non-list inputs, ensuring robust error handling.

2. **Boundary Test Cases**:  
   - The test suite also includes edge cases, such as:
     - **Empty list**:  
       ```python
       assert filter_odd_numbers([]) == []
       ```  
       Verifies the function handles the smallest valid input size.
     - **Negative integers**:  
       ```python
       assert filter_odd_numbers([-1, -2, -3, -4, -5]) == [-1, -3, -5]
       ```  
       Ensures negative integers are processed correctly as odd numbers.
     - **Mixed types** (e.g., floats, strings, `None`):  
       ```python
       assert filter_odd_numbers([1, 2, "three", 4.5, None]) == [1]
       ```  
       Confirms non-integer inputs are ignored without errors.
     - **Single integer input**:  
       ```python
       assert filter_odd_numbers([1]) == [1]
       ```  
       Tests the smallest non-empty valid input.
     - **Zero in the list**:  
       ```python
       assert filter_odd_numbers([0, 1, 2]) == [1]
       ```  
       Verifies that zero is not considered odd and is ignored.
     - **All floats**:  
       ```python
       assert filter_odd_numbers([1.1, 2.2, 3.3]) == []
       ```  
       Confirms floats are excluded entirely.

I’ll work on simplifying the code further and will also review the tests to ensure all possible scenarios are covered. If you have any additional suggestions or specific improvements in mind, feel free to share. Thanks again for your helpful feedback!"

**Review by MadiMalik — CHANGES_REQUESTED** [review]
Thanks, Shahroz, for improving your test file! Now you just need to simplify the implementation of your function to meet the following checks:
  - Variables are named using snake_case.
  - Variable names are clear and descriptive.
  - The code follows the chosen strategy in the simplest way possible.
  - The implementation is as straightforward as possible, given the strategy.

**Comment by Shahroz657** [comment]
Hey Madiha, 

I just updated the files with the requested changes. Thanks for the constructive feedback. Please feel free to review.

**Review by MadiMalik — APPROVED** [review]
Great job, Shahroz! Thank you for improving the implementation of your code

**Review by MadiMalik — APPROVED** [review]

**Review by MadiMalik — APPROVED** [review]

**Review by MadiMalik — APPROVED** [review]

**Review by MadiMalik — APPROVED** [review]

**Comment by MadiMalik** [comment]
I forgot to mention: to link this solution to the corresponding issue, make a small change to the code and push it. Ensure that the commit message ends with “resolve #45” so that the issue will be automatically closed once the changes are merged
